Emby.Server.Implementations/Syncplay/SyncplayController.cs

```diff
@@ -95,131 +102,167 @@ private void CheckDisposed()
         }
 
+        /// <summary>
+        /// Builds a new playback command with some default values.
+        /// </summary>
+        /// <param name="type">The command type.</param>
+        /// <value>The SendCommand.</value>
         private SendCommand NewSyncplayCommand(SendCommandType type)
         {
-            var command = new SendCommand();
-            command.GroupId = _group.GroupId.ToString();
-            command.Command = type;
-            command.PositionTicks = _group.PositionTicks;
-            command.When = _group.LastActivity.ToUniversalTime().ToString("o");
-            command.EmittedAt = DateTime.UtcNow.ToUniversalTime().ToString("o");
-            return command;
+            return new SendCommand()
+            {
+                GroupId = _group.GroupId.ToString(),
+                Command = type,
+                PositionTicks = _group.PositionTicks,
+                When = DateToUTCString(_group.LastActivity),
+                EmittedAt = DateToUTCString(DateTime.UtcNow)
+            };
         }
 
+        /// <summary>
+        /// Builds a new group update message.
+        /// </summary>
+        /// <param name="type">The update type.</param>
+        /// <param name="data">The data to send.</param>
+        /// <value>The GroupUpdate.</value>
         private GroupUpdate<T> NewSyncplayGroupUpdate<T>(GroupUpdateType type, T data)
         {
-            var command = new GroupUpdate<T>();
-            command.GroupId = _group.GroupId.ToString();
-            command.Type = type;
-            command.Data = data;
-            return command;
+            return new GroupUpdate<T>()
+            {
+                GroupId = _group.GroupId.ToString(),
+                Type = type,
+                Data = data
+            };
         }
 
         /// <inheritdoc />
         public void InitGroup(SessionInfo session)
         {
             _group.AddSession(session);
-            _syncplayManager.MapSessionToGroup(session, this);
+            _syncplayManager.AddSessionToGroup(session, this);
 
             _group.PlayingItem = session.FullNowPlayingItem;
             _group.IsPaused = true;
             _group.PositionTicks = session.PlayState.PositionTicks ??= 0;
             _group.LastActivity = DateTime.UtcNow;
 
-            var updateSession = NewSyncplayGroupUpdate(GroupUpdateType.GroupJoined, DateTime.UtcNow.ToUniversalTime().ToString("o"));
-            SendGroupUpdate(session, BroadcastType.SingleSession, updateSession);
+            var updateSession = NewSyncplayGroupUpdate(GroupUpdateType.GroupJoined, DateToUTCString(DateTime.UtcNow));
+            SendGroupUpdate(session, BroadcastType.CurrentSession, updateSession, CancellationToken.None);
             var pauseCommand = NewSyncplayCommand(SendCommandType.Pause);
-            SendCommand(session, BroadcastType.SingleSession, pauseCommand);
+            SendCommand(session, BroadcastType.CurrentSession, pauseCommand, CancellationToken.None);
```

Comment:
Good question. I guess it depends on how locking happens underneath on the WebSocket. If I'm not wrong, one message at a time can be sent across the connection, so it's a matter of which message gets sent first (reception will match the same order).

I'll look into it, I was assuming this FIFO queue behaviour when calling `SendGroupUpdate` or `SendCommand`.

Reply:
> I was assuming this FIFO queue behaviour

I figured as much from the code, so I decided to raise a question earlier :) Finding root causes of bugs which have like 0.5% chance to happen is _hard_, so it's usually better to be extra-cautious in coding and reviewing :)

Reply:
The [docs](https://docs.microsoft.com/en-us/dotnet/api/system.net.websockets.websocket.sendasync?view=netcore-3.1#System_Net_WebSockets_WebSocket_SendAsync_System_ArraySegment_System_Byte__System_Net_WebSockets_WebSocketMessageType_System_Boolean_System_Threading_CancellationToken_) say that: `Exactly one send and one receive is supported on each WebSocket object in parallel.`.

By looking at `.NET`'s [code](https://github.com/dotnet/corefx/blob/release/3.1/src/Common/src/System/Net/WebSockets/ManagedWebSocket.cs) I found the `ManagedWebSocket` class that implements the `WebSocket` abstract class:
```
[ManagedWebSocket class]
Thread-safety:
    - It's acceptable to call ReceiveAsync and SendAsync in parallel.  One of each may run concurrently.
    - It's acceptable to have a pending ReceiveAsync while CloseOutputAsync or CloseAsync is called.
    - Attempting to invoke any other operations in parallel may corrupt the instance.  Attempting to invoke
      a send operation while another is in progress or a receive operation while another is in progress will
      result in an exception.
```
So it looks like calling `SendAsync` from _multiple_ threads could result in an exception being thrown.

In Jellyfin, a single `WebSocket` is wrapped and managed by the `SharpWebSocket` class, which has no locking mechanism implemented.

I'm surprised that everything _seems_ to work and no exception is thrown... :thinking: I guess that's because the WebSockets aren't used that often in other parts of the server.

`ManagedWebSocket` has some internal lock but I wouldn't rely on it, mainly because I don't see it mentioned in the docs. I guess it would be better for `SharpWebSocket` to implement a FIFO queue for outgoing messages.

Another solution would be to await for every single call to `SendGroupUpdate` or `SendCommand`, but that's ugly.

[This](https://www.codetinkerer.com/2018/06/05/aspnet-core-websockets.html) article looks helpful, mainly points 4, 5, 6 and 10. They talk about thread-safety, handling disconnections and Keepalives (apparently all the new code that handles them can be replaced by adding 3 lines in `Jellyfin.Serber/Startup.cs` when calling `app.UseWebSockets()` :sweat_smile:).

Reply:
> I'm surprised that everything _seems_ to work and no exception is thrown... 🤔 I guess that's because the WebSockets aren't used that often in other parts of the server.

We actually have some weird exceptions related to websockets logged time and again... right @Bond-009? So this might explain those.